### PR TITLE
SAML2 Service Provider Pull Request

### PR DESCRIPTION
**Pull request by cmrockwell**
The intent of this PR is to get it reviewed and provide me with valuable feedback about making it better, and ideally getting direct support from Sling Developers. 

Work left to do
1. Sync attributes released by the IDP 
2. Create a metadata provider servlet
3. Consider whether use of `SAML2ConfigService` and `SAML2ConfigServiceImpl` is a good design or not. 
* Pros: Having configs as a service allows Saml2UserMgtServiceImpl and 
AuthenticationHandlerSAML2 to reference the same configuration. 
* Cons: Other bundles installed could access the JKS information configured. 
@Reference
private SAML2ConfigService saml2ConfigService;
4. If o.a.s.auth.form exported TokenStore and SessionStorage then these could be removed. But as it stands for this Whiteboard PR that is not currently the case.

My assumption about the purpose of a Sling Whiteboard is it is similar to a whiteboard in an office. An area to sketch ideas. Stuff gets merged into the Whiteboard that dies there, and maybe a few things get promoted. If this gets promoted, then I assume the questions listed above and a lot more would need answers.

https://issues.apache.org/jira/browse/SLING-9397

### CODE REVIEW

File: saml-handler/README.md [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+  - Sync user membership of groups as defined in the OSGi configs
+- Packages
+  - `idp` is a test fixture based on the OpenSAML V3 eBook. It will be useful for minimizing 
+  setup for testing purposes. Set to disabled for production.  
+  - `sp` is the package for service provider classes, servlets and the 
+  Sling authentication handler
+  - `Helpers` static utilities for help using the opensaml library
+    
+ 
+This code is Derivative Works from [webprofile-ref-project-v3](https://bitbucket.org/srasmusson/webprofile-ref-project-v3)
```
Comment (rombert) [inline saml-handler/README.md:27]: Thank you for the note. I am not sure if we need something else, maybe a `NOTICE` file? Something to be discussed on the dev list.

File: saml-handler/pom.xml [hunk 2]
```diff
@@ -0,0 +1,378 @@
...
+    <sling.password>admin</sling.password>
+  </properties>
+
+  <packaging>bundle</packaging>
+
+  <build>
+    <plugins>
+      <plugin>
+        <groupId>org.apache.felix</groupId>
+        <artifactId>maven-bundle-plugin</artifactId>
```
Comment (rombert) [inline saml-handler/pom.xml:42]: I would suggest moving to the bnd-maven-plugin, as all of Sling is doing that and it makes it simpler for us to maintain and evolve the module.

File: saml-handler/pom.xml [hunk 3]
```diff
@@ -0,0 +1,378 @@
...
+        <extensions>true</extensions>
+        <configuration>
+          <instructions>
+            <Bundle-Activator>org.apache.sling.auth.saml2.Activator</Bundle-Activator>
+            <Embed-Dependency>*;scope=compile|runtime;</Embed-Dependency>
+            <Embed-Transitive>true</Embed-Transitive>
+            <Embed-Directory>target/dependency</Embed-Directory>
+            <Embed-StripGroup>true</Embed-StripGroup>
+            <Import-Package>
+              !sun.misc;resolution:=optional,
```
Comment (rombert) [inline saml-handler/pom.xml:53]: It would be good to note where all of these imports are coming from. I think you're probably embedding too much and need to manually trim the imports. Have you tried using the `Conditional-Package` bnd instruction?

File: saml-handler/pom.xml [hunk 4]
```diff
@@ -0,0 +1,378 @@
...
+      </plugin>
+
+      <plugin>
+        <groupId>org.apache.maven.plugins</groupId>
+        <artifactId>maven-jar-plugin</artifactId>
+        <version>2.6</version>
+        <configuration>
+          <archive>
+            <manifestFile>${project.build.outputDirectory}/META-INF/MANIFEST.MF</manifestFile>
+            <manifestEntries>
```
Comment (rombert) [inline saml-handler/pom.xml:97]: Have you considered adding these entries to the bnd.bnd or pom.xml config of the maven-bundle-plugin?

File: saml-handler/pom.xml [hunk 5]
```diff
@@ -0,0 +1,378 @@
...
+              <Jaas-ModuleClass>org.apache.sling.auth.saml2.sync.Saml2LoginModule</Jaas-ModuleClass>
+            </manifestEntries>
+          </archive>
+        </configuration>
+      </plugin>
+
+      <plugin>
+        <groupId>org.apache.sling</groupId>
+        <artifactId>maven-sling-plugin</artifactId>
+        <version>2.4.2</version>
```
Comment (rombert) [inline saml-handler/pom.xml:107]: Why do you need to the plugin versions? They should be inherited from the parent pom.

File: saml-handler/pom.xml [hunk 6]
```diff
@@ -0,0 +1,378 @@
...
+          <slingUrl>http://${sling.host}:${sling.port}/system/console</slingUrl>
+          <user>${sling.user}</user>
+          <password>${sling.password}</password>
+        </configuration>
+      </plugin>
+
+      <plugin>
+        <groupId>org.apache.rat</groupId>
+        <artifactId>apache-rat-plugin</artifactId>
+        <version>0.13</version>
```
Comment (rombert) [inline saml-handler/pom.xml:118]: Why do you need to the plugin versions? They should be inherited from the parent pom.

File: saml-handler/pom.xml [hunk 7]
```diff
@@ -0,0 +1,378 @@
...
+          </archive>
+        </configuration>
+      </plugin>
+
+      <plugin>
+        <groupId>org.apache.sling</groupId>
+        <artifactId>maven-sling-plugin</artifactId>
+        <version>2.4.2</version>
+        <configuration>
+          <slingUrl>http://${sling.host}:${sling.port}/system/console</slingUrl>
```
Comment (rombert) [inline saml-handler/pom.xml:109]: These kinds of settings should be available out-of-the-box from the maven-sling-plugin.

File: saml-handler/pom.xml [hunk 8]
```diff
@@ -0,0 +1,378 @@
...
+        </configuration>
+      </plugin>
+    </plugins> 
+  </build>
+
+  <dependencies>
+    <dependency>
+      <groupId>junit</groupId>
+      <artifactId>junit</artifactId>
+      <version>4.12</version>
```
Comment (rombert) [inline saml-handler/pom.xml:133]: You should not need to define the version, it's inherited from the parent pom.

File: saml-handler/pom.xml [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+      <version>3.5</version>
+      <scope>provided</scope>
+    </dependency>
+
+<!--    OSGI Annotations-->
+    <dependency>
+      <groupId>org.osgi</groupId>
+      <artifactId>org.osgi.core</artifactId>
+      <version>6.0.0</version>
+      <scope>provided</scope>
```
Comment (rombert) [inline saml-handler/pom.xml:321]: This version and scope should be inherited from the parent pom.

File: saml-handler/pom.xml [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+    </dependency>
+    <dependency>
+      <groupId>org.osgi</groupId>
+      <artifactId>osgi.cmpn</artifactId>
+      <version>6.0.0</version>
```
Comment (rombert) [inline saml-handler/pom.xml:326]: This version and scope should be inherited from the parent pom.

File: saml-handler/pom.xml [hunk 11]
```diff
@@ -0,0 +1,378 @@
...
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.osgi</groupId>
+      <artifactId>osgi.annotation</artifactId>
+      <version>6.0.1</version>
```
Comment (rombert) [inline saml-handler/pom.xml:332]: This version and scope should be inherited from the parent pom.

File: saml-handler/pom.xml [hunk 12]
```diff
@@ -0,0 +1,378 @@
...
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.osgi</groupId>
+      <artifactId>org.osgi.service.component.annotations</artifactId>
+      <version>1.3.0</version>
```
Comment (rombert) [inline saml-handler/pom.xml:338]: This version and scope should be inherited from the parent pom.

File: saml-handler/pom.xml [hunk 13]
```diff
@@ -0,0 +1,378 @@
...
+      <scope>provided</scope>
+    </dependency>
+    <dependency>
+      <groupId>org.osgi</groupId>
+      <artifactId>org.osgi.service.metatype.annotations</artifactId>
```
Comment (rombert) [inline saml-handler/pom.xml:343]: This version and scope should be inherited from the parent pom.

File: saml-handler/pom.xml [hunk 14]
```diff
@@ -0,0 +1,378 @@
...
+      <version>1.3.0</version>
+      <scope>provided</scope>
+    </dependency>
+
+    <dependency>
+      <groupId>javax.jcr</groupId>
+      <artifactId>jcr</artifactId>
+      <version>2.0</version>
```
Comment (rombert) [inline saml-handler/pom.xml:351]: This version and scope should be inherited from the parent pom.

File: saml-handler/src/main/java/org/apache/sling/auth/saml2/AuthenticationHandlerSAML2.java [hunk 15]
```diff
@@ -0,0 +1,661 @@
...
+@Component(
+        service = AuthenticationHandler.class ,
+        name = AuthenticationHandlerSAML2.SERVICE_NAME,
+        configurationPid = "org.apache.sling.auth.saml2.impl.SAML2ConfigServiceImpl",
+        configurationPolicy = ConfigurationPolicy.REQUIRE,
+        property = {"sling.servlet.methods={GET, POST}",
+            AuthenticationHandler.PATH_PROPERTY+"={}",
+            AuthenticationHandler.TYPE_PROPERTY + "=" + AuthenticationHandlerSAML2.AUTH_TYPE,
+            "service.description=SAML2 Authentication Handler",
+            "service.ranking=42",
```
Comment (rombert) [inline saml-handler/src/main/java/org/apache/sling/auth/saml2/AuthenticationHandlerSAML2.java:90]: Any particular reason for this service.ranking value :-) ?

File: saml-handler/src/main/java/org/apache/sling/auth/saml2/AuthenticationHandlerSAML2.java [hunk 16]
```diff
@@ -0,0 +1,661 @@
...
+        },
+        immediate = true)
```
Comment (rombert) [inline saml-handler/src/main/java/org/apache/sling/auth/saml2/AuthenticationHandlerSAML2.java:92]: Why do you need this set to immediate?

**Comment by cmrockwell** [comment]
Thanks for the feedback. I will try to evaluate and respond to your comments about the setup soon. 

FYI I'm just now noticing an issue with the tokens and login is not working. So i am trying to analyze this problem which I did not have yesterday...

> 01.04.2020 17:36:02.377 *ERROR* [qtp1839482286-753] org.apache.sling.auth.saml2.TokenStore AuthNCookie value '5772724430d430ef90fe1b5ad3838515042e1c6f@1585791362376@bob' has expired 999985600001ms ago

Meanwhile, I have added a JKS to the file system under sling/keys. It could be used for Jetty https, but I also plan on using this for signing, verifying and encryption/decryption credentials.

File: saml-handler/src/main/java/org/apache/sling/auth/saml2/AuthenticationHandlerSAML2.java (hunk 15 above)
Comment (cmrockwell) [reply saml-handler/src/main/java/org/apache/sling/auth/saml2/AuthenticationHandlerSAML2.java:91]: haha sort of, yeah. When I first started, I could not get Sling to use this authHandler. I tried high and low numbers, but only the form handler would get used. it seems only path length really has an impact in terms of which authentication handler 'wins' Since the service ranking seemed ineffective, I set it to a low but unique number between 0 and 500 lol

Comment (kwin) [reply saml-handler/src/main/java/org/apache/sling/auth/saml2/AuthenticationHandlerSAML2.java:91]: `service.ranking` is supposed to be an int! If it is a String is it not effective!

File: saml-handler/pom.xml [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+    <plugins>
+      <plugin>
+        <groupId>org.apache.felix</groupId>
+        <artifactId>maven-bundle-plugin</artifactId>
+        <version>3.5.0</version>
+        <extensions>true</extensions>
+        <configuration>
+          <instructions>
+            <Bundle-Activator>org.apache.sling.auth.saml2.Activator</Bundle-Activator>
+            <Embed-Dependency>*;scope=compile|runtime;</Embed-Dependency>
```
Comment (rombert) [inline saml-handler/pom.xml:48]: You don't need to do this _now_, but the bundle is very heavy right now. It's at 18MB in size and embeds 40 jars, including some that should be available as OSGi bundles, e.g. XML parsers, commons-*, guava, metrics. That should be reviewed at some point.

Are the opensaml jars OSGi-aware or is there the potential to make them OSGi-aware?

File: saml-handler/pom.xml [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+              org.apache.oro.text.perl.*;resolution:=optional,
+              antlr.*;resolution:=optional,
+              org.apache.tools.ant.*;resolution:=optional,
+              junit.framework.*;resolution:=optional,
+              org.dom4j.*;resolution:=optional,
+              org.jdom.*;resolution:=optional,
+              org.bouncycastle.*;resolution:=optional,
+              com.sun.org.apache.xerces.internal.*;resolution:=optional,
+              *</Import-Package>
+            <_noee>true</_noee>
```
Comment (rombert) [inline saml-handler/pom.xml:71]: Why is this needed?

File: saml-handler/src/main/java/org/apache/sling/auth/saml2/AuthenticationHandlerSAML2.java (hunk 15 above)
Comment (cmrockwell) [reply saml-handler/src/main/java/org/apache/sling/auth/saml2/AuthenticationHandlerSAML2.java:91]: Awesome! thanks for the feedback. That makes a big difference, so I think service.ranking should be in the configurations.

File: saml-handler/pom.xml (hunk 18 above)
Comment (cmrockwell) [reply saml-handler/pom.xml:71]: If I remove the `_noee` directive, I get the error below and the bundle doesn't activate. I'll take a look at your plugin related questions and circle back to this.

06.04.2020 15:51:23.218 *ERROR* [Background Update org.apache.sling.auth.saml2 (179)] org.apache.felix.http.jetty Cannot install or update bundle from /var/folders/ww/vd006cgx4zgdvvxcfk02b5xr0000gp/T/install2386711195585220288.tmp (org.osgi.framework.BundleException: Unable to resolve org.apache.sling.auth.saml2 [179](R 179.190): missing requirement [org.apache.sling.auth.saml2 [179](R 179.190)] osgi.ee; (&(osgi.ee=JavaSE)(version=9.0)) Unresolved requirements: [[org.apache.sling.auth.saml2 [179](R 179.190)] osgi.ee; (&(osgi.ee=JavaSE)(version=9.0))])
org.osgi.framework.BundleException: Unable to resolve org.apache.sling.auth.saml2 [179](R 179.190): missing requirement [org.apache.sling.auth.saml2 [179](R 179.190)] osgi.ee; (&(osgi.ee=JavaSE)(version=9.0)) Unresolved requirements: [[org.apache.sling.auth.saml2 [179](R 179.190)] osgi.ee; (&(osgi.ee=JavaSE)(version=9.0))]
	at org.apache.felix.framework.Felix.resolveBundleRevision(Felix.java:4133)
	at org.apache.felix.framework.Felix.startBundle(Felix.java:2118)
	at org.apache.felix.framework.BundleImpl.start(BundleImpl.java:998)
	at org.apache.felix.webconsole.internal.core.BaseUpdateInstallHelper.run(BaseUpdateInstallHelper.java:192)
	at java.lang.Thread.run(Thread.java:748)

File: saml-handler/pom.xml (hunk 5 above)
Comment (cmrockwell) [reply saml-handler/pom.xml]: I removed the version

File: saml-handler/src/main/java/org/apache/sling/auth/saml2/AuthenticationHandlerSAML2.java (hunk 16 above)
Comment (cmrockwell) [reply saml-handler/src/main/java/org/apache/sling/auth/saml2/AuthenticationHandlerSAML2.java]: If not set to immediate, then after uploading it to /system/console/bundles it is in the 'Installed' state requiring the user to click the Start button to activate the bundle. Either way is fine. I figure if someone is going to upload or install this bundle, then they might want it Active.

File: saml-handler/pom.xml (hunk 17 above)
Comment (cmrockwell) [reply saml-handler/pom.xml:51]: I tried timing it down a bit, but still 16.6MB. Seems the first priority is to go with bnd plugin, and I would really appreciate any help help with that I could get. BND seems to have a lot of features that could help reduce the size.

File: saml-handler/pom.xml (hunk 2 above)
Comment (cmrockwell) [reply saml-handler/pom.xml]: I made the change to bnd-maven-plugin

File: saml-handler/pom.xml (hunk 3 above)
Comment (cmrockwell) [reply saml-handler/pom.xml]: I could not figure out how to get this bundle to build using Conditional-Package. I made the change to bnd at least.

File: saml-handler/pom.xml (hunk 4 above)
Comment (cmrockwell) [reply saml-handler/pom.xml:81]: I tried taking this suggestion using bnd-maven-plugin without maven-jar-plugin. The doc https://github.com/bndtools/bnd/tree/master/maven/bnd-maven-plugin suggest it is typical to use maven-jar-plugin. I tried removing it but could not make it result in a OSGI bundle. If this is important please provide a little bit more detail. Thanks

File: saml-handler/pom.xml (hunk 17 above)
Comment (cmrockwell) [reply saml-handler/pom.xml]: It is a little bit less now at 11MB still heavy. openSAML is not designed for OSGI. If this module advances, I hope to maybe engage the Shibboleth team to create an osgi friendly version in the future, but that is not their priority.

File: saml-handler/pom.xml (hunk 18 above)
Comment (cmrockwell) [reply saml-handler/pom.xml]: After changing to bnd-maven-plugin the noee is gone and everything seems to work with the new bundle.

File: saml-handler/README.md (hunk 1 above)
Comment (cmrockwell) [reply saml-handler/README.md]: I added a notice file.

**Comment by rombert** [comment]
I've been thinking some more about how to make sure the reviews are productive and reduce the time needed to get this into the whiteboard.

The idea I came up with takes two complementary approaches:

1. Simplify testing
1. Reduce the amount of submitted code

For item 1 I suggest that you provide a docker script or docker-compose setup that launches a SAML-enable identity provider. One idea would be [Keycloak with Docker](https://www.keycloak.org/getting-started/getting-started-docker), but I admit I'm not at all familiar with identity providers to offer an informed suggestion.

This would allow you to drop ~400 LOC in the `idp` package and maybe some other parts that are only used from there.

For item 2 item 1 already helps :-) I think you can start with submitting the minimal functionality that works - and I get that is the AuthenticationHandler.

I see code for the ExternalIdentityProvider and LoginModule as well. If that is not needed for the basic login flow, I would suggest submitting them as follow-up PRs once we merge the initial one.

I also see a potential of dropping some code with the TokenStore class. You mention it's derived from the class in the `o.a.s.auth.form` bundle. Maybe that's something we can export, or you can inline the class in this bundle. The code looks quite complex and is large, I think it's a good idea to keep the duplication out of the bundle.

Would that work for you?

**Comment by cmrockwell** [comment]
Having a builtin IDP during development was really useful and simplified the setup and testing. I get the point is that there should be another way to test the SP code without the demo IDP. I'm curious whether you tried to test it with the internal IDP. This would be the simplest way.

I thought the mock IDP might potentially help in troubleshooting issues, which is why I considered an enable/disable switch for it. Nevertheless, the internal IDP and Saml2ExternalIdentityProvider can be removed from the PR. It wont work without Saml2LoginModule as it stands today, but let me know what code changes would allow it to work without a login module.

I will keep the internal IDP on a dev branch, because it is so handy. But otherwise could remove some code from the PR. At the same time I still need to add code for a few features. 

1. User attribute synchronization: allows setting of user properties like email, given and family names, address, etc based on the data contained in the Assertion.

2. SP metadata provider servlet: this should simply IDP configuration since most IDP's allow 
SP's to be registered by inputing the metadata. Without this admins might need to craft the data by hand

There is going to be some code to review because there's a lot involved with implementing a SAML2 Service Provider authentication provider. Do you have a target number for LOC?

**Comment by rombert** [comment]
@cmrockwell - if you think the IDP is useful, let's keep it in for now. I am trying to make my review simpler, but if that makes you PR submission harder it does not make sense.

**Comment by cmrockwell** [comment]
Thanks, that's kind Robert. 
The demo IDP doesn't offer much if any value to end users. If it creates confusion, uncertainty or concern; then there is no reason to push it. As a developer, I can push it as a separate bundle to my localhost if I want. 
The demo IDP has been removed, so please take a look if things look a little cleaner now.
Cris
 
 
  On Tue, Apr 14, 2020 at 3:20 AM, Robert Munteanu<notifications@github.com> wrote:   


@cmrockwell - if you think the IDP is useful, let's keep it in for now. I am trying to make my review simpler, but if that makes you PR submission harder it does not make sense.

**Comment by klcodanr** [comment]
@cmrockwell @rombert I wonder if we want to merge this PR and then further refine with additional commits / PRs?

**Comment by rombert** [comment]
@klcodanr  - we would need at least a "legal" review in terms of code attribution and ownership. But I agree, we should not polish code indefinitely, but instead be more agile with reviewing.

**Comment by cmrockwell** [comment]
I've added attribution and the original license from webprofile-ref-project-v3. I've also reached out to Stefan Rasmusson to see whether this is sufficient.

https://blog.samlsecurity.com/p/opensaml.html?showComment=1587589973632

**Comment by rombert** [comment]
Thanks  a lot @cmrockwell . I see that the `LICENSE` file is there, so that is good. I am not sure if we need a `NOTICE` file though. I am trying to go through https://www.apache.org/dev/licensing-howto.html#alv2-dep , especially the section titled _Bundling an Apache-2.0-licensed dependency_. 

From an ASF point of view it does not seem that we need to change the NOTICE file, since the third party repository does not have a NOTICE file itself. A notice file should be generated automatically with the build ( but the build fails out-of-the-box for me). BTW, the possible names for the notice are `NOTICE` (preferred) and `NOTICE.txt`.

The remaining question is how and if we need to add attribution to the source code files. Maybe @bdelacretaz or @cziegeler have an idea?

Bertrand, Carsten - @cmrockwell has based some of his submission on code from https://bitbucket.org/srasmusson/webprofile-ref-project-v3, which is Apache-2.0 licensed. The changes are currently listed in https://github.com/apache/sling-whiteboard/blob/8386886dbe241020e11ae8c75ca487a204fe0bbf/saml-handler/NOTICE.md ( although I consider we don't need anything in the NOTICE file, see above ). How would you recommend that we record this attribution?

---

@cmrockwell - once clarified I think we can merge this and then iterate. I suggest that we can keep a record of open items. This can be a simple TODO file in the repository root, a Jira task titled 'SAML authentication handler initial contribution', or anything else really. What would you prefer?

Finally, I have not found an ICLA on file for you, I would encourage you to file on ( see https://www.apache.org/licenses/contributor-agreements.html ) as you seem to be on track for continued contributions.

**Comment by cziegeler** [comment]
As the code is ASL2 and does not require a notice or anything else, we don't need to mention in. But I think its usually good style to do so and have a single sentence in our NOTICE that we include (modified) code from ... which has ASL2 as the license

**Comment by cmrockwell** [comment]
@rombert I updated the NOTICE based on my understanding of Carsen's comment. 
Created SLING-9397 to track the TODO items related to this initial PR.
Also submitted my ICLA.

Please let me know what the build issue is, and what if any git operations were preformed.
Does your ~/.m2/settings.xml have any mirrors? If so you might add an exception for shibboleth libraries.

<mirrorOf>!shibboleth,*</mirrorOf>

**Comment by cmrockwell** [comment]
@rombert 
> but the build fails out-of-the-box for me

Please post more about that

**Comment by rombert** [comment]
@cmrockwell - what I get is

```$ mvn -U clean verify 
[WARNING] SystemTray is not supported, skipping notifications...
[INFO] Scanning for projects...
[INFO] 
[INFO] ------------< org.apache.sling:org.apache.sling.auth.saml2 >------------
[INFO] Building SAML2 Service Provider 0.1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
Downloading from shibboleth: https://build.shibboleth.net/nexus/content/groups/public/org/relaxng/com.springsource.org.relaxng.datatype/1.0.0/com.springsource.org.relaxng.datatype-1.0.0.pom
Downloading from central: https://repo.maven.apache.org/maven2/org/relaxng/com.springsource.org.relaxng.datatype/1.0.0/com.springsource.org.relaxng.datatype-1.0.0.pom
[WARNING] The POM for org.relaxng:com.springsource.org.relaxng.datatype:jar:1.0.0 is missing, no dependency information available
Downloading from shibboleth: https://build.shibboleth.net/nexus/content/groups/public/org/relaxng/com.springsource.org.relaxng.datatype/1.0.0/com.springsource.org.relaxng.datatype-1.0.0.jar
Downloading from central: https://repo.maven.apache.org/maven2/org/relaxng/com.springsource.org.relaxng.datatype/1.0.0/com.springsource.org.relaxng.datatype-1.0.0.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  1.351 s
[INFO] Finished at: 2020-05-04T15:52:37+02:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal on project org.apache.sling.auth.saml2: Could not resolve dependencies for project org.apache.sling:org.apache.sling.auth.saml2:jar:0.1.0-SNAPSHOT: Could not find artifact org.relaxng:com.springsource.org.relaxng.datatype:jar:1.0.0 in shibboleth (https://build.shibboleth.net/nexus/content/groups/public) -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/DependencyResolutionException```

I have no settings.xml. I am running with ```$ mvn -v
Apache Maven 3.6.3 (SUSE 3.6.3-2.1)
Maven home: /usr/share/maven
Java version: 11.0.7, vendor: Oracle Corporation, runtime: /usr/lib64/jvm/java-11-openjdk-11
Default locale: en_US, platform encoding: UTF-8
OS name: "linux", version: "5.6.6-1-default", arch: "amd64", family: "unix"```

At any rate, we'll have to solve the issue of external artifacts at some points, we should not have any custom repositories set.

**Comment by cmrockwell** [comment]
Thanks for posting. Here is what the OpenSAML (Shibboleth) developers say about Maven Central. 
https://wiki.shibboleth.net/confluence/display/DEV/Use+of+Maven+Central

> Maven itself has no support for validating signatures of artifacts (be they signed jars or jars with a detached PGP signature).... If other people aren't worried about the veracity of the artifacts they use that's on them

**Comment by rombert** [comment]
Ack, I read that. I understand the reasons. Note that at the ASF the primary source of disitribution are cryptographically signed source releases, not binaries on Maven Central.

That does not change the fact that Maven Central discourages using `repositories` and `pluginRepository` elements  - see https://central.sonatype.org/pages/requirements.html . If we do this we expose consumers to instability.

There are ways around it, for now we just need to keep in mind that we (probably) should not add extra Maven repositories and fix it before a first release.

**Comment by cmrockwell** [comment]
Sound good. Thank you for merging this. I'll spend some time to analyze the build issue you identified and report back once I'm able to reproduce it.

**Comment by rombert** [comment]
The build works for me with that change. Can you submit a PR with it?
